File: racesdk/java-shims/ShimsJava/src/main/java/ShimsJava/ConnectionType.java
//
// Copyright 2023 Two Six Technologies
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package ShimsJava;

import java.util.HashMap;
import java.util.Map;

public enum ConnectionType {
    CT_UNDEF(0), // undefined
    CT_DIRECT(1), // direct
    CT_INDIRECT(2), // indirect
    CT_MIXED(3), // mixed
    CT_LOCAL(4); // local

    private int value = 0;
    private static Map map = new HashMap<>();

    private ConnectionType(int value) {
        this.value = value;
    }

    static {
        for (ConnectionType connectionType : ConnectionType.values()) {
            map.put(connectionType.value, connectionType);
        }
    }

    public static ConnectionType valueOf(int connectionType) {
        return (ConnectionType) map.get(connectionType);
    }

    public int getValue() {
        return value;
    }
}

File: racesdk/common/source/ConnectionType.cpp

//
// Copyright 2023 Two Six Technologies
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

#include "ConnectionType.h"

#include <stdexcept>

std::string connectionTypeToString(ConnectionType connectionType) {
    switch (connectionType) {
        case CT_UNDEF:
            return "CT_UNDEF";
        case CT_DIRECT:
            return "CT_DIRECT";
        case CT_INDIRECT:
            return "CT_INDIRECT";
        case CT_MIXED:
            return "CT_MIXED";
        case CT_LOCAL:
            return "CT_LOCAL";
        default:
            return "ERROR: INVALID CONNECTION TYPE: " + std::to_string(connectionType);
    }
}

ConnectionType connectionTypeFromString(const std::string &connectionTypeString) {
    if (connectionTypeString == "CT_UNDEF") {
        return CT_UNDEF;
    } else if (connectionTypeString == "CT_DIRECT") {
        return CT_DIRECT;
    } else if (connectionTypeString == "CT_INDIRECT") {
        return CT_INDIRECT;
    } else if (connectionTypeString == "CT_MIXED") {
        return CT_MIXED;
    } else if (connectionTypeString == "CT_LOCAL") {
        return CT_LOCAL;
    } else {
        throw std::invalid_argument("Invalid argument to connectionTypeFromString: " +
                                    connectionTypeString);
    }
}

std::ostream &operator<<(std::ostream &out, ConnectionType connectionType) {
    return out << connectionTypeToString(connectionType);
}